### ISSUE TITLE
Kiali uses a lot of CPU (in validations)

### ISSUE THREAD

**Issue by rolandkool** [kiali/kiali#8007]
### Describe the bug

On a cluster with around 1600 namespaces, but a discovery selector that only selects around 380 of them and cluster wide view set to true, the CPU usage of the Kiali pod is permanently around 2500m. Although I have no reference to compare it with, this seems a bit excessive. This is even with any clients interacting with the UI.

![image](https://github.com/user-attachments/assets/7acaf779-3d2e-4291-8fa0-9586d63a27ab)

The graph shows a drop in usage. With cluster wide view disabled, CPU load is around 4000m.

### Expected Behavior

I'd like to understand if this a reasonable amount of CPU for our setup or if there is an issue with Kiali that causes it to use so much CPU.

### What are the steps to reproduce this bug?

Nothing in particular.
Cluster has around 1600 namespaces in total, but only 380 are selected with a discovery selector. Cluster wide view is enabled. Anonymous mode enabled.

### Environment

- **Kiali version:** 2.2.0
- **Istio version:** 1.23.2 asm 2
- **Kubernetes impl:** GKE
- **Kubernetes version:** 1.30
- **Other notable environmental factors:**

Issues/PRs for this Epic:
- [x] https://github.com/kiali/kiali/pull/8119
- [x] https://github.com/kiali/kiali/pull/8194
- [x] https://github.com/kiali/kiali/pull/8206
- [x] #8211 
- [x] #8210 
- [ ] [pre-compute validation keys for istio config objects](https://github.com/kiali/kiali/pull/8335)

**Comment by rolandkool** [comment]
[profile.gz](https://github.com/user-attachments/files/18168277/profile.gz)

I've attached a 15 seconds CPU profile.

**Comment by rolandkool** [comment]
Deployed another instance on a cluster with similar amount of total namespaces, but only 25 are included in the discovery selector. CPU usage is permanently around 600m.

**Comment by jshaughn** [comment]
I think currently the Kiali Discovery Selector is gating the namespaces available to, or can be seen, by the user.  But I don't think it has the ability to reduce the number of k8s watches that are in place in cluster-wide, to manage caching. And so I think that is why you see more of a constant consumption based on the total namespaces.  

I think the workaround is to not use cluster-wide, despite the fact that it is often the more efficient option.  I think we will need to investigate whether we can apply Discovery Selectors to also minimize cache impact.

cc @jmazzitelli , @nrfox

**Comment by rolandkool** [comment]
@jshaughn thanks for your answer. I tried without cluster-wide access, but that makes the CPU usage even higher (see graph in description above. 4 cpu is without cluster wide access, 2.5 is with).

**Comment by jshaughn** [comment]
OK @rolandkool , thanks for the update. I will add this to the backlog to see what we may be able to do.

**Comment by frittentheke** [comment]
@jshaughn  I am no expert in hunting down performance issues using pprof... but by peeking into the profile @rolandkool  provided it seems that all of that CPU time is spent somewhere in https://github.com/kiali/kiali/blob/master/business/istio_validations.go:

![Image](https://github.com/user-attachments/assets/d5810e86-1bb9-4501-8107-83fdcfc95b05)


@rolandkool  do you happen to have more details on e.g. the reconcile time (see https://github.com/kiali/kiali/blob/cf8785962ffd76bfa861df35a3de8d102fdd2e5f/controller/validations.go#L106)

**Comment by hhovsepy** [comment]
This is the Istio configs validations cache reconciliation, which starts in a background every 10 seconds, does the validation of all Istio config objects and sets it into the cache to be reused by Kiali. The caching mechanism was applied for the Kiali pages loading time optimization.

**Comment by jshaughn** [comment]
Thanks, @frittentheke .  We will have to take a closer look at this, and see what we can do.

**Comment by frittentheke** [comment]
There also are all those `/metrics` exported by Kiali ... if @rolandkool was collecting them, that might also narrow down where Kiali was spending most of its time.

**Comment by jshaughn** [comment]
cc @nrfox 

@hhovsepy , I understand why we want to cache validation results; doing them on-demand creates UI lag for the user, and adding the new controller was a great UX perf enhancement. But I am questioning the frequency of recalculation.  In general, I don't think configuration frequently changes, and so every 10s seems quite aggressive. Of course it's very safe, and prevents obsolete validations, but it's having a noticeable effect on the Kiali pod (and possibly the cost of running Kiali).

I know we have talked about a watcher-based approach, that watches for config changes and reconciles validations only in response to a change.  Is this a feasible approach at this time?  Otherwise, I think we need to cache validations at Kiali startup and then only after a configurable period.  And the default period should likely be much greater than 10s.  Also, it doesn't have to be "all or nothing", we could use a timer and also add watchers incrementally.

**Comment by hhovsepy** [comment]
Note for future work: We need to make sure that Istio configurations edited outside of the Kiali are also included into validations rerun.

**Comment by rolandkool** [comment]
@frittentheke 
From the debug logging, I see this:

> 2025-02-04T15:17:54Z DBG [ValidationsReconciler] Started reconciling 
2025-02-04T15:19:34Z DBG [ValidationsReconciler] Finished reconciling in 100469ms
2025-02-04T15:19:34Z WRN [ValidationsReconciler] Reconcile took longer than the reconcile interval of [10s]. If this continues, validations will be increasingly stale. Please ensure your environment is healthy. If the issue still persists, please open an issue at www.github.com/kiali/kiali
2025-02-04T15:19:34Z DBG [ValidationsReconciler] Started reconciling 
2025-02-04T15:21:15Z DBG [ValidationsReconciler] Finished reconciling in 100436ms
2025-02-04T15:21:15Z WRN [ValidationsReconciler] Reconcile took longer than the reconcile interval of [10s]. If this continues, validations will be increasingly stale. Please ensure your environment is healthy. If the issue still persists, please open an issue at www.github.com/kiali/kiali
2025-02-04T15:21:15Z DBG [ValidationsReconciler] Started reconciling 

So it seems indeed that it is in a constant reconcile loop.
It takes about a 100 seconds to do a single reconcile (on the larger cluster), and when it has only 10s to do so, it starts a new one right away. On the smaller cluster, a reconcile takes 2 seconds. The load is around 500m on that cluster.

Feels like this 10s is a bit too aggressive. The configuration is not that volatile. It would be nice if it would only reconcile on an event or otherwise just allow a different reconcile interval. For example, 1 hour is probably sufficient for us. Most of our users are interested in the graph and dependencies between services.

Are there any metrics from the `/metrics` endpoint you are interested in?

**Comment by jshaughn** [comment]
Thanks @rolandkool.  Right, the fact that the reconcile is longer than the interval is not good.  Also, I'm a little curious about what we're doing that takes so long to calculate, we may want to do some detailed profiling of that validation logic.  I think event-based reconcile would be best, but otherwise we need a configurable intervale, with a much longer default.  Also, the next reconcile could depend on the length of the previous reconcile duration.

**Comment by nrfox** [comment]
> Some possible interesting metrics (from a Kiali instance that has been running for about 7 minutes):

This is great data, thank you!

> Feels like this 10s is a bit too aggressive. The configuration is not that volatile. It would be nice if it would only reconcile on an event or otherwise just allow a different reconcile interval. For example, 1 hour is probably sufficient for us. Most of our users are interested in the graph and dependencies between services.

Yes we should provide a config option to tune and/or disable this. The reason it was set so low to begin with was to ensure that when you make a config change, you get relatively fast feedback on whether or not that config change was valid. But there's also other tools you can use to validate your config, like `istioctl analyze`, so it's perfectly reasonable to set this higher and/or disable it altogether.

Moving to event based reconciling would help overall reconcile times after the initial validation. There's an underlying perf issue with validations that still needs to be addressed though. Moving validations to background processing was a way of improving the user experience but it did not deal with the underlying issue. Before we moved validations to being processed in the background, the validations requests would sometimes takes 1-2 minutes which would lead to significant slowdowns in the UI. It's now a better experience but obviously there's now the issue of Kiali constantly consuming resources if validations are slow.

**Comment by hhovsepy** [comment]
To summarize the solution directions, we can create the following sub-issues to resolve and implement:

- [ ]  Make the reconciliation interval configurable and/or provide an option to disable it.
- [ ] Implement event-based reconciliation triggered by configuration changes outside of Kiali.
- [ ] Optimize the validation process to reduce resource usage.

**Comment by nrfox** [comment]
Yeah and in terms of level of effort:

This one is very trivial
> Make the reconciliation interval configurable and/or provide an option to disable it.

This one is a medium-large effort
> Implement event-based reconciliation triggered by configuration changes outside of Kiali.

This one requires some investigation
> Optimize the validation process to reduce resource usage.

IMO we should prioritize 1 and 3.

**Comment by jshaughn** [comment]
I agree with the prioritization of 1 and 3.  Let's get the setting configurable asap and change the default to what, 60m?  Also, don't start the timer until after a reconcile completes.

Do we reconcile when Kiali itself is changing/updating config?

After that change is in, let's prioritize the 3rd task, try to understand if we can speed up the reconcile.

**Comment by hhovsepy** [comment]
> 
> Do we reconcile when Kiali itself is changing/updating config?
> 
Yeas each time Kiali edits a config, it is reconciled.

**Comment by nrfox** [comment]
> Let's get the setting configurable asap and change the default to what, 60m?

60m is very large. Validations will be noticeably stale at that point. `60s`? We'll keep the warning so that if you start to go past that interval then you'll at least get a warning message telling you to adjust the config option.

**Comment by rolandkool** [comment]
Thanks, looking forward for the fix. That will allow me to roll out Kiali across more clusters. This high usage was holding me back.
I can help testing a pre release if you like.

**Comment by frittentheke** [comment]
@nrfox @jshaughn could you possible compile a few instructions on how to do e.g. pprof profiling that actually would help you narrow down this issue? Certainly it makes sense to also (!) tackle the underlying (potential) performance issues.
Years ago I also was involved in https://github.com/kiali/kiali/issues/4592 which was (in the end) quite easily found when looking at the profiling in the real environment.

**Comment by nrfox** [comment]
@frittentheke you can have a look at https://kiali.io/docs/configuration/debugging-kiali/#profiler for how to enable the profiler.

**Comment by frittentheke** [comment]
> [@frittentheke](https://github.com/frittentheke) you can have a look at https://kiali.io/docs/configuration/debugging-kiali/#profiler for how to enable the profiler.

@rolandkool As I don't see similar durations as you do, could you potentially run some more pprof ?

**Comment by rolandkool** [comment]
[profile.gz](https://github.com/user-attachments/files/18672578/profile.gz)

Here's another 15 seconds profile.

The profile command on the /debug/pprof page does not work, you need to manually add `&seconds=<nr>` to the URL.

**Comment by nrfox** [comment]
@rolandkool as of v2.6 you can set `external_services.istio.validation_reconcile_interval` to have more control over how often Kiali will perform validations: https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.istio.validation_reconcile_interval. You can see if that helps with the CPU usage.

**Comment by jshaughn** [comment]
I'm going to take a look, I'm unfamiliar with the validation logic, so it may take some time...

**Comment by rolandkool** [comment]
@nrfox thanks for implementing this setting. That definitely helps in our setup. I've set the reconcile interval to 15m, which is good enough for us for now. When not reconciling, the load drops to a more manageable 50-100m (depending on whether the UI is used or not).

Looking forward to a rework of this logic so that it only processes delta's instead of reconciling the whole config each time.

**Comment by frittentheke** [comment]
@jshaughn we don't observe the issue, but if you need any more data (pprof) let me know. I suppose @rolandkool is just as willing to provide any debug data you require to figure out the suspected inefficiency here.

**Comment by jshaughn** [comment]
@frittentheke , thanks, I may accept the offer. At the moment, though, I am just familiarizing myself with the code and doing some initial re-arranging.  I'll let you know!

**Comment by nrfox** [comment]
> Reducing the number of allocations we do in validations should help with this.

That being said, if this is going to take a large scale refactor, it would be nice to at least investigate this issue first: https://github.com/kiali/kiali/issues/4713 since the results of that may also require a refactor.

**Comment by jshaughn** [comment]
> Reducing the number of allocations we do in validations should help with this.

Is that using the old code or the new code?

**Comment by frittentheke** [comment]
> > Reducing the number of allocations we do in validations should help with this.
> 
> That being said, if this is going to take a large scale refactor, it would be nice to at least investigate this issue first: https://github.com/kiali/kiali/issues/4713 since the results of that may also require a refactor.

+1

I was always wondering why Kiali did his own tests and not built on top of istio analyse (even contribute missing consistency and housekeeping checks there).

**Comment by nrfox** [comment]
> Is that using the old code or the new code?

This was looking at the profile linked in this issue but I've seen similar results when running a local benchmark: https://github.com/kiali/kiali/pull/8208

**Comment by frittentheke** [comment]
> > I was always wondering why Kiali did his own tests and not built on top of istio analyse (even contribute missing consistency and housekeeping checks there).
> 
> iirc Kiali validations predate istio analyze but istio analyze has been around for awhile now

thanks for https://github.com/kiali/kiali/issues/4713#issuecomment-2698846877 @nrfox

**Comment by jshaughn** [comment]
> If you look at the cpu profile a lot of time seems to be spent on garbage collection:

@nrfox The work for #8223 should help a lot with this.  See the pprof results that @hhovsepy has produced.

**Comment by jshaughn** [comment]
@frittentheke , @rolandkool , Kiali 2.7 will contain several changes that we hope will provide perf benefits, particularly for
validation.  It is scheduled for release Monday, Mar 17 (Happy St Patrick's Day!)  When/if you get the chance, please let us know if you are seeing improvement.

**Comment by rolandkool** [comment]
@jshaughn I've just deployed 2.7.1 on our clusters. With a reconcile interval of 15 minutes, I do see that each reconcile takes less time (up to a minute less). Still heavy on CPU though.

**Comment by jshaughn** [comment]
Thanks for the feedback, @rolandkool. I would have expected the first reconcile to be a bit faster, but assuming no config changes, I would have expected subsequent reconciles to be improved a fair amount.  This is because we added some logic to detect configuration changes.  We'll keep looking for ways to improve.

**Comment by rolandkool** [comment]
Here's a comparison.
On 2.6:

![Image](https://github.com/user-attachments/assets/b7ca4303-eafb-4db9-a94f-27b5e59b4dd3)

On 2.7.1:

![Image](https://github.com/user-attachments/assets/92a37545-68db-48bb-ab37-2266f1bd47aa)

**Comment by jmazzitelli** [comment]
Should Kial provide a on/off switch for validation entirely? For those environments that simply can't support it without killing resources, just have people turn it off entirely.

**Comment by jshaughn** [comment]
> Should Kial provide a on/off switch for validation entirely? 

I don't think so. Validations are a valuable tool and the reconcile interval is configurable.  But, if requested we could do something like have a reconcile interval <= 0 be a disable.

**Comment by jshaughn** [comment]
This round of validation perf enhancements is complete. It can still be intensive but we've made significant reductions to CPU and memory. We will also allow disabling validation in https://github.com/kiali/kiali/issues/8317.
